auth/auth.go

```diff
@@ -412,115 +411,100 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
-// issuer in the token with a provider.
-// Used to authenticate a JWT token coming from an insecure source (e.g. client request)
-// If the token is validated but the user for the username defined in the subject claim doesn't exist,
-// creates the user when autoRegister=true.
-func (auth *Authenticator) AuthenticateUntrustedJWT(token string, providers OIDCProviderMap, callbackURLFunc OIDCCallbackURLFunc) (User, jose.JWT, error) {
-
-	base.Debugf(base.KeyAuth, "AuthenticateJWT called with token: %s", base.UD(token))
+// Authenticates a user based on a JWT token string and a set of providers. Attempts to match the issuer in the
+// token with a provider. Used to authenticate a JWT token coming from an insecure source (e.g. client request).
+// If the token is validated but the user for the username defined in the subject claim doesn't exist, creates
+// the user when autoRegister=true.
+func (auth *Authenticator) AuthenticateUntrustedJWT(token string, providers OIDCProviderMap, callbackURLFunc OIDCCallbackURLFunc) (User, *oidc.IDToken, error) {
+	base.Debugf(base.KeyAuth, "AuthenticateUntrustedJWT called with token: %s", base.UD(token))
 
-	// Parse JWT (needed to determine issuer/provider)
-	jwt, err := jose.ParseJWT(token)
+	// Get the ID Token from the raw JWT string; determine issuer/provider.
+	idToken, err := GetIDToken(token)
 	if err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing JWT in AuthenticateJWT: %v", err)
-		return nil, jose.JWT{}, err
-	}
-
-	// Get client for issuer
-	issuer, audiences, err := GetJWTIssuer(jwt)
-	base.Debugf(base.KeyAuth, "JWT issuer: %v, audiences: %v", base.UD(issuer), base.UD(audiences))
-	if err != nil {
-		base.Debugf(base.KeyAuth, "Error getting JWT issuer: %v", err)
-		return nil, jose.JWT{}, err
+		base.Debugf(base.KeyAuth, "Error parsing JWT string to IDToken in AuthenticateUntrustedJWT: %v", err)
+		return nil, &oidc.IDToken{}, err
 	}
 
 	base.Debugf(base.KeyAuth, "Call GetProviderForIssuer w/ providers: %+v", base.UD(providers))
-	provider := providers.GetProviderForIssuer(issuer, audiences)
+	provider := providers.GetProviderForIssuer(idToken.Issuer, idToken.Audience)
 	base.Debugf(base.KeyAuth, "Provider for issuer: %+v", base.UD(provider))
-
 	if provider == nil {
-		return nil, jose.JWT{}, base.RedactErrorf("No provider found for issuer %v", base.UD(issuer))
+		return nil, &oidc.IDToken{}, base.RedactErrorf("No provider found for issuer %v", base.UD(idToken.Issuer))
 	}
 
 	// VerifyJWT validates the claims and signature on the JWT
 	client := provider.GetClient(callbackURLFunc)
 	if client == nil {
-		return nil, jose.JWT{}, fmt.Errorf("OIDC client was not initialized")
+		return nil, &oidc.IDToken{}, fmt.Errorf("OIDC client was not initialized")
 	}
 
-	err = client.VerifyJWT(jwt)
+	config := &oidc.Config{ClientID: *provider.ClientID}
+	idTokenVerifier := provider.OIDCClient.Provider.Verifier(config)
```

Comment:
As discussed, we'd prefer to create a Verifier once and cache it for use for multiple authentications, as was previously done with the OIDCClient.  Is there a reason that can't be done?